### PR TITLE
Fixed a crashed caused by parsing functional pseudo as simple selector

### PR DESCRIPTION
**Pull request by sjbanov**
Added checks to selector parsing function.

Should I add example from https://github.com/dart-lang/tools/issues/1168?
```
.foo {
  border-radius: calc(0 - 1px);
}
```
I don't understand it usage.

### CODE REVIEW

File: lib/parser.dart [hunk 1]
```diff
@@ -179,6 +179,21 @@ final _legacyPseudoElements = Set<String>.from(const [
   'first-line',
 ]);
 
+// CSS3 functional pseudo-classes (https://www.w3.org/TR/2018/REC-selectors-3-20181106/#selectors)
```
Comment (kevmoo) [inline lib/parser.dart:182]: Maybe bump the min SDK to 2.2 and use an actual `Set` constant?

File: lib/parser.dart [hunk 2]
```diff
@@ -179,6 +179,21 @@ final _legacyPseudoElements = Set<String>.from(const [
...
+final _functionalPseudoClasses = new Set<String>.from(const [
+  'nth-child',
+  'nth-last-child',
+  'nth-of-type',
+  'nth-last-of-type',
+  'lang',
+  'not',
+]);
+
+final _compoundSelectors = new Set<String>.from(const [
```
Comment (kevmoo) [inline lib/parser.dart]: ditto

File: lib/parser.dart (hunk 1 above)
Comment (sjbanov) [reply lib/parser.dart:182]: Fixed

File: lib/parser.dart (hunk 2 above)
Comment (sjbanov) [reply lib/parser.dart]: Fixed

File: lib/parser.dart [hunk 3]
```diff
@@ -179,6 +179,21 @@ final _legacyPseudoElements = Set<String>.from(const [
   'first-line',
 ]);
 
+// CSS3 functional pseudo-classes (https://www.w3.org/TR/2018/REC-selectors-3-20181106/#selectors)
+final _functionalPseudoClasses = {
```
Comment (kevmoo) [inline lib/parser.dart]: you can make these const, too!

**Comment by leonsenft** [comment]
Would you mind adding a regression test for the `calc(0 - 1px)` case? I realize it looks nonsensical, but I'm guessing it's the result of a higher level style language (like Sass) compiling a repeated pattern down and the zero on the left hand side is just one iteration of a loop.

**Review by leonsenft — CHANGES_REQUESTED** [review]
Unfortunately I think this solution might be too brittle and inflexible. Keeping an exhaustive list of allowed functional pseudo-classes and pseudo-elements means this package is always at risk of breaking. While it's probably feasible to update the package when a new pseudo-class or -element is standardized, I don't think tracking all of the vendor-prefixed ones is reasonable.

I understand this fixes an immediate problem for some clients, but I'm afraid it would introduce more problems for others (specifically for vendor-prefixed pseudos).

**Comment by sjbanov** [comment]
@leonsenft , Please check the new attempt to correct the bug. dart-lang/csslib#94

**Pull request closed without merging**